common/libs/framing/frame_detector.h

```diff
@@ -0,0 +1,129 @@
+#ifndef __FRAME_DETECTOR__
+#define __FRAME_DETECTOR__
+
+#include "serial_listeners.h"
+#include <stdint.h>
+#include <string.h>
+
+template <class RxBuffer, int FRAME_BUF_LEN>
+class FrameDetector : public RxListener {
+public:
+  enum class State {
+    LOST,       // Frame detector does not know where it is in the stream of
+                // bytes coming into RxBuffer
+    WAIT_START, // Frame detector for a frame start marker
+    RX_FRAME    // Frame detector is receiving a frame
+  };
+
+  FrameDetector(RxBuffer &t) : rx_buffer_(t){};
+
+  // Starts frame detector
+  [[nodiscard]] bool Begin() {
+    state_ = State::LOST;
+    return rx_buffer_.Begin(this);
+  }
+
+  // Callback method called when RxBuffer is full
+  void onRxComplete() override {
+    // We should never reach the full read of rx buffer.
+    // If we get here, this means, there are no marker
+    // chars in the stream, so we are lost
+    error_counter_++;
+    state_ = State::LOST;
+    rx_buffer_.RestartRX(this);
+  }
+
+  // Callback method called when we receive a marker character
+  void onCharacterMatch() override {
+    switch (state_) {
+    case State::LOST:
+      // if we have received something before this marker,
+      // we assume, this is the frame end marker, so wait
+      // for start
+      if (rx_buffer_.ReceivedLength() > 1) {
+        state_ = State::WAIT_START;
+        // if we were lucky to get lost in the interframe silence,
+        // assume this is the start of the frame
+      } else if (rx_buffer_.ReceivedLength() == 1) {
+        state_ = State::RX_FRAME;
+      } else {
+        // TODO alert, safe reset
+        // Should never end up here
+        // DMA is not working?
+      }
+      break;
+    case State::WAIT_START:
+      if (rx_buffer_.ReceivedLength() == 1) {
+        state_ = State::RX_FRAME;
+      } else {
+        // some junk received while waiting for start marker,
+        // but should have been just silence
+        error_counter_++;
+        state_ = State::LOST;
+      }
+      break;
+    case State::RX_FRAME:
+      // end marker received, check if we got something
+      if (rx_buffer_.ReceivedLength() > 1) {
+        processReceivedData();
+        state_ = State::WAIT_START;
+      } else {
+        // repeated marker char received
+        // assume we are still good
```

Comment:
What does "still good" mean?

Reply:
We have received an extra frame marker just after receiving a marker we have decided to start a frame on.
```
bbbMMMffff...
    ^ go to RECEIVING_FRAME
     ^ this is an extra marker we received
       ^ assume we will receive a proper frame after this     
```

Reply:
What we can do here:
 * Just decode 0 length frame - does not make much sense, but maybe sender got crazy and is sending 0 length frames
 * go to LOST - maybe
 * ignore and continue decoding, it this MARK was caused by a bit flip in stream, upper level decoder will catch that

I will add more tests and see if it creates any problems.

Reply:
So far I think this implementation is correct as it allows us to recover in many situations related to callbacks racing.